File: 046/cpp/main.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        if (nums.empty()) {
            return {{}};
        }
        vector<vector<int>> res;
        for (int i = 0; i < nums.size(); i++) {
            int t = nums[i];
            vector<int> nums0(nums);
            nums0.erase(nums0.begin() + i);
            auto res0 = permute(nums0);
            for (auto &r0: res0) {
                r0.insert(r0.begin(), t);
                res.push_back(r0);
            }
        }
        return res;
    }
};

File: 046/rust/main.rs
impl Solution {
    pub fn permute0(nums: &Vec<i32>) -> Vec<Vec<i32>> {
        let n = nums.len();
        if n == 0 {
            return vec![Vec::new()];
        }

        let mut res = Vec::new();
        for i in 0..n {
            let t = nums[i];
            let mut nums0 = nums.clone();
            nums0.remove(i);
            let res0 = Solution::permute0(&nums0);
            for r0 in res0 {
                let mut r = r0.clone();
                r.insert(0, t);
                res.push(r);
            }
        }
        return res;
    }

    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        return Solution::permute0(&nums);
    }
}
